ggml-aarch64.cpp

```diff
@@ -0,0 +1,1856 @@
+// SPDX-FileCopyrightText: Copyright 2024 Arm Ltd.
+
+#pragma GCC diagnostic ignored "-Wpedantic"
+#pragma GCC diagnostic ignored "-Wignored-attributes"
+
+#define GGML_COMMON_IMPL_C
+#include "ggml-common.h"
+
+#include "ggml-quants.h"
+#include "ggml-impl.h"
+
+#define GGML_COMMON_IMPL_C
+#include "ggml-common.h"
+
+#include <math.h>
+#include <string.h>
+#include <assert.h>
+#include <float.h>
+#include <stdlib.h> // for qsort
+#include <stdio.h>  // for GGML_ASSERT
+
+#include "ggml-aarch64.h"
+
+#define UNUSED GGML_UNUSED
+
+size_t quantize_q4_0_aarch64(const float * GGML_RESTRICT src, void * GGML_RESTRICT dst, int64_t nrow, int64_t n_per_row, const float * quant_weights) {
+    if (!quant_weights) {
+        int nrows_interleaved = 1;
+        int blocklen_per_row;
+
+#if defined(__ARM_FEATURE_SVE)
+        if (svcntw() == 8) {
+            nrows_interleaved = 8;
+            blocklen_per_row = 8;
+        }
+        else if (ggml_cpu_has_neon() && ggml_cpu_has_matmul_int8()) {
+            nrows_interleaved = 4;
+            blocklen_per_row = 8;
+        }
+#elif defined(__ARM_NEON) && defined(__ARM_FEATURE_MATMUL_INT8)
+        nrows_interleaved = 4;
+        blocklen_per_row = 8;
+#elif defined(__ARM_NEON)
+        nrows_interleaved = 4;
+        blocklen_per_row = 4;
+#endif
```

Comment:
This would make the quantization order hardware specific?

Reply:
We fixed it in the latest commit. The functionality of the quantization function is no longer hardware-dependent. To make the quantization function non-hardware specific, we split our single consolidated interleaved layout type Q4_0_AARC64 type into three separate types: Q4_0_4_4, Q4_0_4_8, and Q4_0_8_8 (depending on the number of rows interleaved and block length per row).

A user can convert Q4_0 to one of the interleaved layout types Q4_0_4_4, Q4_0_4_8, and Q4_0_8_8 on any hardware, and then run the appropriate format based on the underlying cpu type.